File: tests/framework/protocol.py
"""
COBS Protocol Implementation

Consistent Overhead Byte Stuffing (COBS) encoding/decoding and CRC-8 calculation.
"""

from typing import List, Optional


def crc8(data: bytes, poly: int = 0x07, init: int = 0x00) -> int:
    """
    Calculate CRC-8 checksum.
    
    Args:
        data: Input bytes
        poly: Polynomial (default 0x07)
        init: Initial value (default 0x00)
    
    Returns:
        CRC-8 checksum byte
    """
    crc = init
    for byte in data:
        crc ^= byte
        for _ in range(8):
            if crc & 0x80:
                crc = ((crc << 1) ^ poly) & 0xFF
            else:
                crc = (crc << 1) & 0xFF
    return crc


def cobs_encode(data: bytes) -> bytes:
    """
    COBS encode data.
    
    Replaces all 0x00 bytes with distance codes, prepends overhead byte.
    Result contains no 0x00 bytes (except optional delimiter).
    
    Args:
        data: Raw data to encode
    
    Returns:
        COBS encoded bytes (without 0x00 delimiter)
    """
    if len(data) == 0:
        return bytes([0x01])
    
    output = bytearray()
    code_idx = 0
    code = 1
    output.append(0)  # Placeholder for first code
    
    for byte in data:
        if byte == 0:
            output[code_idx] = code
            code_idx = len(output)
            output.append(0)  # Placeholder for next code
            code = 1
        else:
            output.append(byte)
            code += 1
            if code == 0xFF:
                output[code_idx] = code
                code_idx = len(output)
                output.append(0)
                code = 1
    
    output[code_idx] = code
    return bytes(output)


def cobs_decode(data: bytes) -> Optional[bytes]:
    """
    COBS decode data.
    
    Args:
        data: COBS encoded bytes (without 0x00 delimiter)
    
    Returns:
        Decoded bytes, or None if invalid
    """
    if len(data) == 0:
        return None
    
    output = bytearray()
    idx = 0
    
    while idx < len(data):
        code = data[idx]
        if code == 0:
            return None  # Invalid: 0x00 in encoded data
        
        idx += 1
        
        for i in range(1, code):
            if idx >= len(data):
                return None  # Truncated
            output.append(data[idx])
            idx += 1
        
        if code < 0xFF and idx < len(data):
            output.append(0)
    
    return bytes(output)


def build_packet(packet_type: int, payload: bytes = b'', tag: int = 0) -> bytes:
    """
    Build a complete COBS-encoded packet.
    
    Packet structure (before COBS):
        [type:u8][tag:u8][len:u16LE][payload:0-512][crc8:u8]
    
    Args:
        packet_type: Packet type byte
        payload: Payload bytes (0-512)
        tag: Request/response correlation tag (0 = async/unsolicited)
    
    Returns:
        COBS encoded packet with 0x00 delimiter
    """
    if len(payload) > 512:
        raise ValueError(f"Payload too large: {len(payload)} > 512")
    
    length = len(payload)
    raw = bytes([packet_type, tag & 0xFF, length & 0xFF, (length >> 8) & 0xFF]) + payload
    raw += bytes([crc8(raw)])
    
    encoded = cobs_encode(raw)
    return encoded + b'\x00'


def parse_packet(data: bytes) -> Optional[tuple]:
    """
    Parse a COBS-encoded packet.
    
    Args:
        data: COBS encoded packet (with or without 0x00 delimiter)
    
    Returns:
        Tuple of (packet_type, tag, payload) or None if invalid
    """
    # Remove delimiter if present
    if data.endswith(b'\x00'):
        data = data[:-1]
    
    decoded = cobs_decode(data)
    if decoded is None or len(decoded) < 5:
        return None
    
    packet_type = decoded[0]
    tag = decoded[1]
    length = decoded[2] | (decoded[3] << 8)  # u16LE
    
    if len(decoded) != 5 + length:
        return None  # Length mismatch
    
    payload = decoded[4:4+length]
    received_crc = decoded[-1]
    
    # Verify CRC
    expected_crc = crc8(decoded[:-1])
    if received_crc != expected_crc:
        return None  # CRC mismatch
    
    return (packet_type, tag, payload)


# Payload helper functions
def u16_le(value: int) -> bytes:
    """Pack unsigned 16-bit little-endian."""
    return bytes([value & 0xFF, (value >> 8) & 0xFF])


def i16_le(value: int) -> bytes:
    """Pack signed 16-bit little-endian."""
    if value < 0:
        value = value + 0x10000
    return u16_le(value)


def u32_le(value: int) -> bytes:
    """Pack unsigned 32-bit little-endian."""
    return bytes([
        value & 0xFF,
        (value >> 8) & 0xFF,
        (value >> 16) & 0xFF,
        (value >> 24) & 0xFF
    ])


def read_u16_le(data: bytes, offset: int = 0) -> int:
    """Read unsigned 16-bit little-endian."""
    return data[offset] | (data[offset + 1] << 8)


def read_i16_le(data: bytes, offset: int = 0) -> int:
    """Read signed 16-bit little-endian."""
    value = read_u16_le(data, offset)
    if value >= 0x8000:
        value -= 0x10000
    return value


def read_u32_le(data: bytes, offset: int = 0) -> int:
    """Read unsigned 32-bit little-endian."""
    return (data[offset] |
            (data[offset + 1] << 8) |
            (data[offset + 2] << 16) |
            (data[offset + 3] << 24))


def crc16_ccitt(data: bytes, init: int = 0xFFFF) -> int:
    """
    Calculate CRC-16/CCITT checksum.

    Matches the C++ StreamProtocol::crc16() implementation.
    Polynomial 0x1021, initial value 0xFFFF.

    Args:
        data: Input bytes
        init: Initial CRC value (default 0xFFFF)

    Returns:
        CRC-16 value (0-65535)
    """
    crc = init
    for byte in data:
        crc ^= byte << 8
        for _ in range(8):
            if crc & 0x8000:
                crc = ((crc << 1) ^ 0x1021) & 0xFFFF
            else:
                crc = (crc << 1) & 0xFFFF
    return crc

File: tools/cli/protocol.go
package main

// ScaleFX CLI - Protocol Layer
// COBS encode/decode, CRC-8/CRC-16, packet build/parse, endian helpers.

// CRC8 calculates CRC-8 checksum (polynomial 0x07).
func CRC8(data []byte) byte {
	var crc byte
	for _, b := range data {
		crc ^= b
		for i := 0; i < 8; i++ {
			if crc&0x80 != 0 {
				crc = (crc << 1) ^ 0x07
			} else {
				crc = crc << 1
			}
		}
	}
	return crc
}

// CRC16CCITT calculates CRC-16/CCITT (polynomial 0x1021, init 0xFFFF).
func CRC16CCITT(data []byte) uint16 {
	crc := uint16(0xFFFF)
	for _, b := range data {
		crc ^= uint16(b) << 8
		for i := 0; i < 8; i++ {
			if crc&0x8000 != 0 {
				crc = (crc << 1) ^ 0x1021
			} else {
				crc = crc << 1
			}
		}
	}
	return crc
}

// COBSEncode encodes data using Consistent Overhead Byte Stuffing.
// Returns encoded bytes without the 0x00 delimiter.
func COBSEncode(data []byte) []byte {
	if len(data) == 0 {
		return []byte{0x01}
	}

	output := make([]byte, 0, len(data)+len(data)/254+2)
	output = append(output, 0) // placeholder for first code
	codeIdx := 0
	code := byte(1)

	for _, b := range data {
		if b == 0 {
			output[codeIdx] = code
			codeIdx = len(output)
			output = append(output, 0) // placeholder for next code
			code = 1
		} else {
			output = append(output, b)
			code++
			if code == 0xFF {
				output[codeIdx] = code
				codeIdx = len(output)
				output = append(output, 0)
				code = 1
			}
		}
	}

	output[codeIdx] = code
	return output
}

// COBSDecode decodes COBS-encoded data. Returns nil if invalid.
func COBSDecode(data []byte) []byte {
	if len(data) == 0 {
		return nil
	}

	output := make([]byte, 0, len(data))
	idx := 0

	for idx < len(data) {
		code := data[idx]
		if code == 0 {
			return nil // invalid: 0x00 in encoded data
		}
		idx++

		for i := byte(1); i < code; i++ {
			if idx >= len(data) {
				return nil // truncated
			}
			output = append(output, data[idx])
			idx++
		}

		if code < 0xFF && idx < len(data) {
			output = append(output, 0)
		}
	}

	return output
}

// BuildPacket builds a complete COBS-encoded packet with 0x00 delimiter.
// Packet structure (pre-COBS): [type:u8][tag:u8][len:u16LE][payload:0-512][crc8:u8]
func BuildPacket(packetType byte, payload []byte, tag byte) []byte {
	length := len(payload)
	raw := make([]byte, 0, 4+length+1)
	raw = append(raw, packetType, tag, byte(length&0xFF), byte((length>>8)&0xFF))
	raw = append(raw, payload...)
	raw = append(raw, CRC8(raw))

	encoded := COBSEncode(raw)
	return append(encoded, 0x00)
}

// ParsePacket decodes a COBS packet and validates CRC.
// Returns (packetType, tag, payload, ok).
func ParsePacket(data []byte) (byte, byte, []byte, bool) {
	// Remove delimiter if present
	if len(data) > 0 && data[len(data)-1] == 0x00 {
		data = data[:len(data)-1]
	}

	decoded := COBSDecode(data)
	if decoded == nil || len(decoded) < 5 {
		return 0, 0, nil, false
	}

	packetType := decoded[0]
	tag := decoded[1]
	length := int(decoded[2]) | int(decoded[3])<<8

	if len(decoded) != 5+length {
		return 0, 0, nil, false // length mismatch
	}

	payload := decoded[4 : 4+length]
	receivedCRC := decoded[len(decoded)-1]
	expectedCRC := CRC8(decoded[:len(decoded)-1])

	if receivedCRC != expectedCRC {
		return 0, 0, nil, false // CRC mismatch
	}

	return packetType, tag, payload, true
}

// --- Endian helpers ---

func U16LE(v uint16) []byte {
	return []byte{byte(v & 0xFF), byte((v >> 8) & 0xFF)}
}

func U32LE(v uint32) []byte {
	return []byte{
		byte(v & 0xFF),
		byte((v >> 8) & 0xFF),
		byte((v >> 16) & 0xFF),
		byte((v >> 24) & 0xFF),
	}
}

func ReadU16LE(data []byte, offset int) uint16 {
	return uint16(data[offset]) | uint16(data[offset+1])<<8
}

func ReadI16LE(data []byte, offset int) int16 {
	return int16(ReadU16LE(data, offset))
}

func ReadU32LE(data []byte, offset int) uint32 {
	return uint32(data[offset]) |
		uint32(data[offset+1])<<8 |
		uint32(data[offset+2])<<16 |
		uint32(data[offset+3])<<24
}
